### PR TITLE
503/Using kotlin coroutines instead of custom executors

### PR DESCRIPTION
**Pull request by vlivanov**
Core change: 

interface TasksDataSource {
    suspend fun getTasks(): List<Task>
    suspend fun getTask(taskId: String): Task?
...
}

getTask() and getTasks() methods throw DataNotAvailableException

Unit tests are passing 100%

### CODE REVIEW

**Comment by vlivanov** [comment]
I signed it!

**Comment by vlivanov** [comment]
I signed it!

File: todoapp/build.gradle [hunk 1]
```diff
@@ -26,7 +26,7 @@ buildscript {
         google()
     }
     dependencies {
-        classpath "com.android.tools.build:gradle:$gradlePluginVersion"
+        classpath 'com.android.tools.build:gradle:3.0.1'
```
Comment (JoseAlcerreca) [inline todoapp/build.gradle]: Revert and modify variable

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/statistics/StatisticsPresenter.kt [hunk 2]
```diff
@@ -45,9 +45,9 @@ class StatisticsPresenter(
         // that the app is busy until the response is handled.
         EspressoIdlingResource.increment() // App is busy until further notice
 
-        tasksRepository.getTasks(object : TasksDataSource.LoadTasksCallback {
-            override fun onTasksLoaded(tasks: List<Task>) {
-                // We calculate number of active and completed tasks
+        runBlocking {
```
Comment (dmytrodanylyk) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/statistics/StatisticsPresenter.kt]: It is designed to bridge regular blocking code to libraries that are written in suspending style, to be used in main functions and in tests. You should use `launch`. See https://github.com/dmytrodanylyk/android-architecture/blob/dev-todo-mvp-kotlin-coroutines/todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/tasks/TasksPresenter.kt#L68

Also it's better to provide context explicitly.

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/data/source/local/TasksLocalDataSource.kt [hunk 3]
```diff
@@ -33,43 +37,41 @@ class TasksLocalDataSource private constructor(
      * Note: [TasksDataSource.LoadTasksCallback.onDataNotAvailable] is fired if the database doesn't exist
      * or the table is empty.
      */
-    override fun getTasks(callback: TasksDataSource.LoadTasksCallback) {
-        appExecutors.diskIO.execute {
+    override suspend fun getTasks(): List<Task> {
```
Comment (dmytrodanylyk) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/data/source/local/TasksLocalDataSource.kt]: `async/await` is designed to be used when you want to return `Deferred`. In this case it's better to use `withContext`. 

Also, do not switch to UI thread, it's up to client when to switch context. See https://github.com/dmytrodanylyk/android-architecture/blob/dev-todo-mvp-kotlin-coroutines/todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/data/source/local/TasksLocalDataSource.kt#L36

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/data/source/TasksDataSource.kt [hunk 4]
```diff
@@ -29,39 +29,15 @@ import com.example.android.architecture.blueprints.todoapp.data.Task
  */
 interface TasksDataSource {
```
Comment (dmytrodanylyk) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/data/source/TasksDataSource.kt]: I decided to make all functions in`TasksDataSource` suspended. This is how coroutines are designed to work. (Indicate that function may do intensive operation.) 

See: https://github.com/dmytrodanylyk/android-architecture/blob/dev-todo-mvp-kotlin-coroutines/todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/data/source/TasksDataSource.kt

(Callbacks to be removed)

**Pull request closed without merging**